src/common/activitymonitor.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor-disposable';
+
+import {
+  ISignal, Signal, clearSignalData
+} from 'phosphor-signaling';
+
+
+/**
+ * A class that monitors activity on a signal.
+ */
+export
+class ActivityMonitor<Sender, Args> implements IDisposable {
+  /**
+   * Construct a new activity monitor.
+   */
+  constructor(options: ActivityMonitor.IOptions) {
+    options.signal.connect(this._onSignalFired, this);
+    this._timeout = options.timeout || 1000;
+  }
+
+  /**
+   * A signal emitted when activity has ceased.
+   */
+  get activityStopped(): ISignal<Sender, Args> {
+    return Private.activityStoppedSignal.bind(this);
+  }
+
+  /**
+   * The timeout associated with the monitor, in milliseconds.
+   */
+  get timeout(): number {
+    return this._timeout;
+  }
+  set timeout(value: number) {
+    this._timeout = value;
+  }
+
+  /**
+   * Test whether the monitor has been disposed.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources used by the activity monitor.
+   */
+  dispose(): void {
+    if (this._isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+    clearSignalData(this);
+  }
+
+  /**
+   * A signal handler for the monitored signal.
+   */
+  private _onSignalFired(): void {
```

Comment:
Actually, I'm not so sure about this.  Do we use the args of the *last* signal even though the args could differ every time?